SPI: Stm32  3-wire (half-duplex) support

tbursztyka opened the pull request:
I still need to test this in non-interrupt base mode.
 
One issue left as well: though transactions are going totally fine while hts221 initialization, it seems to hang at some point (sensor_sample_fetch() transaction I believe). Don't know why yet, this needs some more investigations.

Commit:
drivers/spi: Add support for half-duplex mode (3-wire) on stm32

Basically, when configured to run on half-duplex mode, it will switch to
TX and RX relevantly (first it TX a word, then it will RX one).

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/sensor: Fix 3-wire support on HTS221

Address auto-increment code is different depending on the bus it is
wired to.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/sensor: Cleaning HTS221 configuration structure

One left-over attribute that does not belong here anymore.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -288,10 +288,27 @@ static void spi_stm32_shift_m(SPI_TypeDef *spi, struct spi_stm32_data *data)
 		/* NOP */
 	}
 
+	if (data->half_duplex) {
+		LL_SPI_SetTransferDirection(spi, LL_SPI_HALF_DUPLEX_TX);
+	}
+
```
Shouldn't this be done just before exiting the function  and after the RX ?

tbursztyka replied:
We loop until the RX fifo is filled in, just above it. At this stage we can already switch back to TX, but I guess it won't change much if it moved at the end of the function, more readable actually.

tbursztyka replied:
half-duplex seems to be very sensitive to timing: the time you switch from RX to TX, you might end up with garbage in the RX pipe. (basically as long as clock line is on, it will be "receiving").

tbursztyka commented:
Funny how - while working on half-duplex mode (either rx or tx)- I still thought about the logic only existing in full-duplex mode... 
So, this is not going to fly if for each rx byte it's going back to tx in order to send an irrelevant dummy byte...

Will try to find some time to spend on this.

Commit:
drivers/sensor: Adapts ST's stmemsc to SPI 3-wire

Buffer handling in half duplex is different than full duplex as there is
no RX byte counterpart to each TX byte.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/sensor: Fix few stmemsc style issue

Indentation and using generic macro.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
Got it to work, but there is still 2 questions left:

- what if a transaction is only about rx? (0 tx bytes)
- what if a transaction is made of multiple tx/rx ? I am not sure we could handle that case with the buffer sets as of today. And are these type of transactions a relevant use case?


One detail, while testing on hts221:
- cold boot of the board, all works as it should
- sensor still plugged in, a hard reset makes the text boot fail (see error log below). sensor VCC unplugged: no crash (sensor cannot initialize successfully obviously). Actually, plugging out/in the VCC, makes reset working properly. Weird behavior...

```
[00:00:00.003,000] <dbg> HTS221: hts221_read_conversion_data: 
                                 50 03 e9 d4 00 00 cd 02  00 00 00 00 00 00 00 00 |P....... ........
*** Booting Zephyr OS build zephyr-v3.3.0-5095-gf20d58dec348 ***
[00:00:00.004,000] <err> os: ***** USAGE FAULT *****
[00:00:00.004,000] <err> os:   Illegal use of the EPSR
[00:00:00.004,000] <err> os: r0/a1:  0x20001c74  r1/a2:  0xe000ed00  r2/a3:  0x20001c74
[00:00:00.004,000] <err> os: r3/a4:  0x00000000 r12/ip:  0xaaaaaaaa r14/lr:  0x0800b1f9
[00:00:00.004,000] <err> os:  xpsr:  0x60000000
[00:00:00.004,000] <err> os: Faulting instruction address (r15/pc): 0x00000000
[00:00:00.004,000] <err> os: >>> ZEPHYR FATAL ERROR 35: Unknown error on CPU 0
[00:00:00.004,000] <err> os: Current thread: 0x20001608 (sysworkq)
[00:00:00.137,000] <err> os: Halting syste
```

tbursztyka commented:
The conversion issue when resetting mentioned in previous comment seems to be related to electrical mishaps. The sensor daughter card is supposed to accept 5V as input, but when I do: it's not working. plugging/unplugging vcc makes the board rebooting, things like that. 

Anyway, removing the DNM label. I however would like to get an answer on the 2 questions in previous comment about transaction types.

MaureenHelm commented:
@erwango will you please take another look?

erwango commented:
> Anyway, removing the DNM label. I however would like to get an answer on the 2 questions in previous comment about transaction types.

@avisconti Would you be able to answer ?

avisconti reviewed (commented):
LGTM, a part of a small non-blocking comment.
Thanks for the fixes in stmemsc and hts221 drivers.

BTW, looks like you have tested everything on htts221 SPI 3-wires. Is this the case? I can do a quick test as soon as I will find some time on few different devices. I'll approve after your valuable comments.

avisconti commented:
> Got it to work, but there is still 2 questions left:
> 
> * what if a transaction is only about rx? (0 tx bytes)
> * what if a transaction is made of multiple tx/rx ? I am not sure we could handle that case with the buffer sets as of today. And are these type of transactions a relevant use case?

Interesting questions. As long as the addressed SPI device is a motion sensor I think the two types of transactions are not an expected use case (Even in rx case a non-empty tx transaction is always needed to write the register addr to read from. And I cannot think of a case in which a single transaction is made up of multiple rx/tx). When the SPI device is something different I'm not sure what the use case would be.

Does it make sense to you what I'm saying?

erwango commented:
@dgastonochoa As you're focusing on STM32 SPI, would you mind having a look to this PR ?

dgastonochoa commented:
> @dgastonochoa As you're focusing on STM32 SPI, would you mind having a look to this PR ?

I will as soon as possible.

dgastonochoa commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -250,6 +250,71 @@ static int spi_stm32_get_err(SPI_TypeDef *spi)
+			}
+
+			/* Switching to RX */
+			LL_SPI_SetTransferDirection(spi, LL_SPI_HALF_DUPLEX_RX);
```
I think maybe the only thing required here is to let the user do a call to `spi_write` and then another call to `spi_read` (instead of transceive). With that, I think we only need to support to configure the SPI peripheral as full-duplex or half-duplex as a function of the SPI cpnfiguration received, and let the user do 2 separated calls (with the appropriate config.) to write and read.

MaureenHelm commented:
@tbursztyka ping
